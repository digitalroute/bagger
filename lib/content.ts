import { parse as validateMediaType } from 'content-type';
import { createSwaggerDefinition } from './utils/create_swagger_definition';
import { Schema } from '@hapi/joi';
import { ContentObject } from 'openapi3-ts';

// There are no type definitions for json-schema-to-openapi-schema
// eslint-disable-next-line @typescript-eslint/no-var-requires
const toOpenApi = require('json-schema-to-openapi-schema');

export class BaggerInvalidMediaTypeError extends Error {}

export class BaggerDuplicateMediaTypeError extends Error {}

export class BaggerSchemaNotSetForMediaType extends Error {}

interface InternalContent {
  mediaType: string;
  schema: Schema;
}

export interface ContentSchemas {
  [key: string]: Schema;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isJoi(schema: any): schema is Schema {
  return schema.isJoi;
}

export class Content {
  private internals: InternalContent[] = [];

  /**
   * Add a content object.
   * @param mediaType The media type, for example `application/json`
   * @param schema JOI Schema for this media type
   */
  public add(mediaType: string, schema: Schema): Content {
    if (!validateMediaType(mediaType)) {
      throw new BaggerInvalidMediaTypeError();
    }
    if (this.internals.some(internal => internal.mediaType === mediaType)) {
      throw new BaggerDuplicateMediaTypeError();
    }
    this.internals.push({
      mediaType,
      schema
    });
    return this;
  }

  public getSchemas(): ContentSchemas {
    return this.internals.reduce((prev: ContentSchemas, curr: InternalContent): ContentSchemas => {
      prev[curr.mediaType] = curr.schema;
      return prev;
    }, {});
  }

  public compile(): ContentObject {
    const obj = this.internals.reduce((prev: ContentObject, curr: InternalContent): ContentObject => {
      prev[curr.mediaType] = {
        schema: isJoi(curr.schema) ? createSwaggerDefinition(curr.schema) : toOpenApi(curr.schema)
      };
      return prev;
    }, {});
    return obj;
  }
}
