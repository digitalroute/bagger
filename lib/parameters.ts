import { ParameterObject, ParameterStyle, ReferenceObject, ExampleObject } from 'openapi3-ts';
import { Content, ContentSchemas } from './content';
import { Schema } from '@hapi/joi';
import { createSwaggerDefinition } from './utils/create_swagger_definition';

class BaggerContentDefinedForParameterError extends Error {}
class BaggerSchemaDefinedForParameterError extends Error {}

interface ExamplesObject {
  [param: string]: ExampleObject | ReferenceObject;
}

export type ParameterType = 'path' | 'query' | 'cookie' | 'header';

export class BaggerParameter {
  private settings: ParameterObject;
  private _content: Content = new Content();
  private _schema?: Schema;

  public constructor(type: ParameterType, name: string) {
    this.settings = { in: type, name };
  }

  /**
   * Returns the type of this parameter. E.g. where the parameter is located.
   */
  public getType(): ParameterType {
    return this.settings.in;
  }
  
  /**
   * Describe the parameter.
   * @param description the description of the parameter.
   */
  public description(description: string): BaggerParameter {
    this.settings.description = description;
    return this;
  }

  /**
   * Marks if the parameter has to be present or not.
   * @param required if the parameter has to be present or not.
   */
  public required(required: boolean): BaggerParameter {
    this.settings.required = required;
    return this;
  }

  /**
   * Marks a parameer as deprecated if set to `true`
   * @param deprecated
   */
  public deprecated(deprecated: boolean): BaggerParameter {
    this.settings.deprecated = deprecated;
    return this;
  }

  /**
   * Query string parameters may only have a name and no value, like:
   * ```
   * GET /foo?metadata
   * ```
   *
   * This marks if an empty value is allowed or not.
   * @param allowEmptyValue if empty values are allowed.
   */
  public allowEmptyValue(allowEmptyValue: boolean): BaggerParameter {
    this.settings.allowEmptyValue = allowEmptyValue;
    return this;
  }

  /**
   * Parameters containing arrays and objects can be serialized in different ways. Style defines how multiple values are delimited.
   * @param style the style for delimiting parameter. Read more at: https://swagger.io/docs/specification/serialization/
   */
  public style(style: ParameterStyle): BaggerParameter {
    this.settings.style = style;
    return this;
  }

  /**
   * Parameters containing arrays and objects can be serialized in different ways.
   * Explode specifies whether arrays and objects should generate separate parameters for each array item or object property.
   * Read more at: https://swagger.io/docs/specification/serialization/
   * @param explode specifies whether arrays and objects should generate separate parameters for each array item or object property.
   */
  public explode(explode: boolean): BaggerParameter {
    this.settings.explode = explode;
    return this;
  }

  public allowReserved(allowReserved: boolean): BaggerParameter {
    this.settings.allowReserved = allowReserved;
    return this;
  }

  public schema(schema: Schema): BaggerParameter {
    if (Object.keys(this._content.getSchemas()).length > 0) {
      throw new BaggerContentDefinedForParameterError();
    }
    this._schema = schema;
    return this;
  }

  public examples(examples: ExamplesObject): BaggerParameter {
    this.settings.examples = examples;
    return this;
  }

  public addContent(contentType: string, schema: Schema): BaggerParameter {
    if (this.settings.schema) {
      throw new BaggerSchemaDefinedForParameterError();
    }
    this._content.add(contentType, schema);
    return this;
  }

  public getSchemas(): ContentSchemas {
    if (this._schema) {
      return {
        'application/json': this._schema
      };
    }
    return this._content.getSchemas();
  }

  public compile(): ParameterObject {
    if (this._schema) {
      this.settings.schema = {
        'application/json': {
          schema: createSwaggerDefinition(this._schema)
        }
      };
    } else {
      this.settings.content = this._content.compile();
    }
    return this.settings;
  }
}
