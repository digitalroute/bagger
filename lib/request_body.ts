import { Content, ContentSchemas } from './content';
import { Schema } from 'joi';
import { RequestBodyObject } from 'openapi3-ts';

export class BaggerUndefinedSchemaError extends Error {}

export class BaggerRequestBody {
  private _description?: string;
  private _required: boolean = false;
  private _content: Content = new Content();

  /**
   * Set the description of the request body. It can be markdown-formatted.
   * @param description The description of the request body.
   */
  public description(description: string): BaggerRequestBody {
    this._description = description;
    return this;
  }

  /**
   * Request bodies are optional by default. To mark the body as required, use required: `true`.
   * @param required if the request body is required.
   */
  public required(required: boolean): BaggerRequestBody {
    this._required = required;
    return this;
  }

  /**
   * Content describes the content of a body.
   * @param mediaType The media type of the content. Like 'application/json' or 'text/plain'. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types for more information
   * @param schema A `joi` schema describing the content of the body. This can also be used for validating requests in run time by using `.getSchema()`.
   */
  public content(mediaType: string, schema: Schema): BaggerRequestBody {
    this._content.add(mediaType, schema);
    return this;
  }

  /**
   * Get the `joi` schemas that is used to describe the body content. It can be used for validating requests in runtime.
   */
  public getSchemas(): ContentSchemas {
    return this._content.getSchemas();
  }

  /**
   * Creates a compiled object representation of the request body.
   */
  public compile(): RequestBodyObject {
    return {
      description: this._description,
      required: this._required,
      content: this._content.compile()
    };
  }
}
