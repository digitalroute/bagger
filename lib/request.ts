import { BaggerResponse } from './response';
import { BaggerRequestBody } from './request_body';
import { BaggerParameter } from './parameters';
import { SchemaStorage } from './schema_storage';
import {
  OperationObject,
  SecurityRequirementObject,
  RequestBodyObject,
  ExternalDocumentationObject,
  ParameterObject,
  ReferenceObject,
  ResponsesObject,
  CallbacksObject,
  ServerObject,
  PathsObject
} from 'openapi3-ts';

export type Method = 'get' | 'post' | 'patch' | 'put' | 'delete' | 'options' | 'head' | 'trace';

type OperationTypes =
  | string
  | ExternalDocumentationObject
  | ParameterObject[]
  | ReferenceObject[]
  | RequestBodyObject
  | ReferenceObject
  | ResponsesObject
  | CallbacksObject
  | boolean
  | SecurityRequirementObject[]
  | ServerObject;

export class BaggerRequest {
  private path: string;
  private method: Method;
  private schemaStorage: SchemaStorage;

  private operationContext: OperationObject = {
    responses: {}
  };
  public readonly isBagger = true;

  public constructor(path: string, method: Method, schemaStorage: SchemaStorage) {
    this.path = path;
    this.method = method;
    this.schemaStorage = schemaStorage;
  }

  private setInContext(key: keyof OperationObject, value: OperationTypes): BaggerRequest {
    if (!this.operationContext[key]) {
      this.operationContext[key] = value;
    }
    return this;
  }

  public operationId(operationId: string): BaggerRequest {
    return this.setInContext('operationId', operationId);
  }

  public deprecated(deprecated: boolean): BaggerRequest {
    return this.setInContext('deprecated', deprecated);
  }

  public summary(summary: string): BaggerRequest {
    return this.setInContext('summary', summary);
  }

  public addTag(tag: string): BaggerRequest {
    if (!this.operationContext.tags) {
      this.operationContext.tags = [];
    }
    this.operationContext.tags.push(tag);
    return this;
  }

  public addSecurity(scheme: string, scopes: string[] = []): BaggerRequest {
    if (!this.operationContext.security) {
      this.operationContext.security = [];
    }
    const obj: SecurityRequirementObject = {};
    obj[scheme] = scopes;
    this.operationContext.security.push(obj);
    return this;
  }

  public body(requestBody: BaggerRequestBody): BaggerRequest {
    this.schemaStorage.addRequestSchemas(this.path, this.method, requestBody.getSchemas(), 'body');
    return this.setInContext('requestBody', requestBody.compile());
  }

  public addParameter(parameter: BaggerParameter): BaggerRequest {
    if (!this.operationContext.parameters) {
      this.operationContext.parameters = [];
    }
    this.schemaStorage.addRequestSchemas(
      this.path,
      this.method,
      parameter.getSchemas(),
      parameter.getType(),
      parameter.getName()
    );
    this.operationContext.parameters.push(parameter.compile());
    return this;
  }

  public addResponse(response: BaggerResponse): BaggerRequest {
    if (!this.operationContext.responses) {
      this.operationContext.responses = {};
    }

    const responseDefinition = response.compile();

    this.operationContext.responses = {
      ...responseDefinition,
      ...this.operationContext.responses
    };
    return this;
  }

  public compile(): PathsObject {
    return {
      [this.path]: {
        [this.method]: this.operationContext
      }
    };
  }
}
