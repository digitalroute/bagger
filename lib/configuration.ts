import { BaggerRequest } from './request';
import {
  ExternalDocumentationObject,
  ServerObject,
  InfoObject,
  OpenAPIObject,
  PathsObject,
  ComponentsObject
} from 'openapi3-ts';
import { cleanObject } from './utils/clean_object';
import { validateSchema } from './utils/validate_schema';
import {
  BaggerSchemaComponent,
  SchemaComponentObject,
  BaggerSecurityComponent,
  SecuritySchemeComponentObject
} from './component';
import { Schema } from '@hapi/joi';

class BaggerMultipleComponentsWithSameNameFoundError extends Error {}
class BaggerComponentNotFoundError extends Error {}

interface SwaggerConfiguration {
  info: InfoObject;
  servers?: ServerObject[];
  externalDocs?: ExternalDocumentationObject;
}

export class BaggerConfiguration {
  private internalConfiguration: BaggerConfigurationInternal;

  public constructor(internalConfiguration: BaggerConfigurationInternal) {
    this.internalConfiguration = internalConfiguration;
  }

  /**
   * Create the information object in an OpenAPI schema.
   * @param info Defines the information object in an OpenAPI schema
   */
  public info(info: InfoObject): BaggerConfiguration {
    this.internalConfiguration.setInfo(info);
    return this;
  }

  /**
   * Add a server to the configuration.
   * @param server The server object
   */
  public addServer(server: ServerObject): BaggerConfiguration {
    this.internalConfiguration.addServer(server);
    return this;
  }

  /**
   * Define external documentation for the OpenAPI schema.
   * @param externalDocs The external documentation object. Containing an URL.
   */
  public externalDocs(externalDocs: ExternalDocumentationObject): BaggerConfiguration {
    this.internalConfiguration.setExternalDocs(externalDocs);
    return this;
  }
}

interface BaggerComponents {
  schemas: BaggerSchemaComponent[];
  securitySchemes: BaggerSecurityComponent[];
}

export class BaggerConfigurationInternal {
  private requests: BaggerRequest[] = [];
  private components: BaggerComponents = {
    schemas: [],
    securitySchemes: []
  };
  private configuration: SwaggerConfiguration = {
    info: {
      title: '',
      version: ''
    }
  };

  public addRequest(request: BaggerRequest): void {
    this.requests.push(request);
  }

  public addSecurityComponent(component: BaggerSecurityComponent): void {
    this.components.securitySchemes.push(component);
  }

  public addSchemaComponent(component: BaggerSchemaComponent): void {
    this.components.schemas.push(component);
  }

  public getSchemaComponent(name: string): Schema {
    const schemas = this.components.schemas.filter(schema => schema.getName() === name);
    if (schemas.length === 0) {
      throw new BaggerComponentNotFoundError();
    }
    if (schemas.length > 1) {
      throw new BaggerMultipleComponentsWithSameNameFoundError();
    }
    return schemas[0].getSchema() as Schema;
  }

  public setInfo(info: InfoObject): void {
    this.configuration.info = info;
  }

  public addServer(server: ServerObject): void {
    if (!this.configuration.servers) {
      this.configuration.servers = [];
    }
    this.configuration.servers.push(server);
  }

  public setExternalDocs(externalDocs: ExternalDocumentationObject): void {
    this.configuration.externalDocs = externalDocs;
  }

  private mergeRequests(): PathsObject {
    const mergedRequests = this.requests
      .map(request => request.compile())
      .reduce((paths, request) => {
        const requestObject = Object.entries(request)[0];
        const [path, definition] = requestObject;

        paths[path] = {
          ...definition,
          ...paths[path]
        };

        return paths;
      }, {});

    return mergedRequests;
  }

  public compileSchemaComponents(): SchemaComponentObject {
    const unmergedComponents = this.components.schemas.map(component => component.compile());
    return unmergedComponents.reduce((schemas: SchemaComponentObject, component) => {
      schemas[component.name] = component.schema;
      return schemas;
    }, {});
  }

  public compileSecurityComponents(): SecuritySchemeComponentObject {
    const unmergedComponents = this.components.securitySchemes.map(component => component.compile());
    return unmergedComponents.reduce((securitySchemes: SecuritySchemeComponentObject, component) => {
      securitySchemes = {
        ...securitySchemes,
        ...component
      };
      return securitySchemes;
    }, {});
  }

  private compileComponents(): ComponentsObject {
    const schemas = this.compileSchemaComponents();
    const securitySchemes = this.compileSecurityComponents();
    return {
      schemas,
      securitySchemes
    };
  }

  public compile(): OpenAPIObject {
    const paths = this.mergeRequests();

    const components = this.compileComponents();

    const swaggerDefinition = cleanObject({
      ...this.configuration,
      openapi: '3.0.0',
      paths,
      components
    });

    const errors = validateSchema(swaggerDefinition);

    if (errors.length > 0) {
      throw new Error(JSON.stringify(errors));
    }

    return swaggerDefinition;
  }
}
