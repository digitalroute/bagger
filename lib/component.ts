import { Schema } from '@hapi/joi';
import { SchemaObject, ReferenceObject, SecuritySchemeType, SecuritySchemeObject } from 'openapi3-ts';
import { createSwaggerDefinition } from './utils/create_swagger_definition';
import { BaggerConfigurationInternal } from './configuration';

export type ComponentType = 'schemas' | 'securitySchemes';

export interface SchemaComponentObject {
  [schema: string]: SchemaObject | ReferenceObject;
}

export interface SecuritySchemeComponentObject {
  [name: string]: SecuritySchemeObject;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isReference(value: any): value is ReferenceObject {
  return value.$ref && typeof value.$ref === 'string';
}

export class BaggerSecurityComponent {
  private key: string;
  private _name?: string;
  private type: SecuritySchemeType;
  private _scheme?: string;
  private _in?: string;
  private _openIdConnectUrl?: string;
  private _description?: string;

  public constructor(key: string, type: SecuritySchemeType) {
    this.key = key;
    this.type = type;
  }

  public scheme(scheme: string): BaggerSecurityComponent {
    this._scheme = scheme;
    return this;
  }

  public in(location: string): BaggerSecurityComponent {
    this._in = location;
    return this;
  }

  public openIdConnectUrl(openIdConnectUrl: string): BaggerSecurityComponent {
    this._openIdConnectUrl = openIdConnectUrl;
    return this;
  }

  public description(description: string): BaggerSecurityComponent {
    this._description = description;
    return this;
  }

  public name(name: string): BaggerSecurityComponent {
    this._name = name;
    return this;
  }

  public compile(): { [key: string]: SecuritySchemeObject } {
    return {
      [this.key]: {
        type: this.type,
        description: this._description,
        in: this._in,
        scheme: this._scheme,
        openIdConnectUrl: this._openIdConnectUrl,
        name: this._name
      }
    };
  }
}

export class BaggerSchemaComponent {
  private name: string;
  private schema: Schema | ReferenceObject;

  public constructor(name: string, schema: Schema) {
    this.name = name;
    this.schema = schema;
  }

  private compileValue(): SchemaObject | ReferenceObject {
    if (isReference(this.schema)) {
      return this.schema;
    } else {
      return createSwaggerDefinition(this.schema);
    }
  }

  public getName(): string {
    return this.name;
  }

  public getSchema(): Schema | ReferenceObject {
    return this.schema;
  }

  public compile(): { name: string; schema: SchemaObject | ReferenceObject } {
    return {
      name: this.name,
      schema: this.compileValue()
    };
  }
}

export class BaggerComponentAdder {
  private internalConfiguration: BaggerConfigurationInternal;

  public constructor(internalConfiguration: BaggerConfigurationInternal) {
    this.internalConfiguration = internalConfiguration;
  }

  public security(component: BaggerSecurityComponent): BaggerSecurityComponent {
    this.internalConfiguration.addSecurityComponent(component);
    return component;
  }

  /**
   * Add a reusable data model (schema).
   * @param name A unique id that is used to referance the component.
   * @param schema A `hapi/joi` schema that describes the data model.
   */
  public schema(name: string, schema: Schema): BaggerSchemaComponent {
    const component = new BaggerSchemaComponent(name, schema);
    this.internalConfiguration.addSchemaComponent(component);
    return component;
  }
}

export class BaggerComponentGetter {
  private internalConfiguration: BaggerConfigurationInternal;

  public constructor(internalConfiguration: BaggerConfigurationInternal) {
    this.internalConfiguration = internalConfiguration;
  }

  public getSchema(name: string): Schema {
    return this.internalConfiguration.getSchemaComponent(name);
  }
}
