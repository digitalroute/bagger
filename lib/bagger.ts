import { BaggerResponse } from './response';
import { BaggerRequest, Method } from './request';
import { BaggerRequestBody } from './request_body';
import { BaggerConfiguration, BaggerConfigurationInternal } from './configuration';
import { OpenAPIObject, SecuritySchemeType } from 'openapi3-ts';
import { BaggerParameter, ParameterType } from './parameters';
import { BaggerComponentAdder, BaggerComponentGetter, BaggerSecurityComponent } from './component';
import { SchemaStorage, SchemaDefinition } from './schema_storage';

export class Bagger {
  private internalConfiguration = new BaggerConfigurationInternal();
  private schemaStorage = new SchemaStorage();
  private configuration: BaggerConfiguration;
  private componentAdder: BaggerComponentAdder;
  private componentGetter: BaggerComponentGetter;

  public constructor() {
    this.configuration = new BaggerConfiguration(this.internalConfiguration);
    this.componentAdder = new BaggerComponentAdder(this.internalConfiguration);
    this.componentGetter = new BaggerComponentGetter(this.internalConfiguration);
  }

  /**
   * Get the validation object for a specific path and method. The result can be used for validation.
   * This comes from `bagger.parameter()` and `bagger.requestBody()`, and they have to be declared on the request before getting them.
   * @param path The url/path to the request
   * @param method HTTP method like 'GET' or 'PUT'
   * @param contentType
   */
  public getRequestSchema(path: string, method: string, contentType: string = 'application/json'): SchemaDefinition {
    return this.schemaStorage.getRequestSchema(path, method, contentType);
  }

  /**
   * Creates a Request object
   * @param path Path representing the request
   * @param method HTTP method for this path. Multiple methods can be added for a path by calling
   * addRequest for the same path multiple times.
   * @example
   * ```js
   * const bagger = require('.');
   * const joi = require('joi');
   *
   * bagger
   *   .addRequest('/bags', 'get')
   *   .addTag('bags')
   *   .addTag('getters')
   *   .addSecurity('OAuth2')
   *   .addParameter(
   *     bagger
   *       .parameter()
   *       .query('bagSize')
   *       .schema(joi.string().valid('10L', '20L', '30L')
   *       .required(true)
   *   )
   *   .addResponse(
   *     bagger
   *       .response(204)
   *       .description('Fetched bag of given size')
   *   );
   * ```
   */
  public addRequest(path: string, method: Method): BaggerRequest {
    const request = new BaggerRequest(path, method, this.schemaStorage);
    this.internalConfiguration.addRequest(request);
    return request;
  }

  /**
   * Add a reusable component that can be referenced to.
   */
  public addComponent(): BaggerComponentAdder {
    return this.componentAdder;
  }

  /**
   * Get a reusable component.
   */
  public getComponent(): BaggerComponentGetter {
    return this.componentGetter;
  }

  /**
   * Get the configuration object for bagger. This defines the configuration for all endpoints.
   * @example
   * ```js
   * const bagger = require('bagger');
   * bagger
   *   .configure()
   *   .info({
   *     title: 'Bagger API',
   *     version: 'v1',
   *     description: 'Provides resources to building swagger definitions'
   *   })
   *   .addServer({
   *     url: 'https://localhost:3000',
   *     description: 'Local development'
   *   })
   * ...
   * ```
   */
  public configure(): BaggerConfiguration {
    return this.configuration;
  }

  /**
   * Compiles bagger into an OpenAPI schema. This has to be done _after_ all requests has been added to
   * bagger.
   * @returns An OpenAPI schema that can be used to render a UI.
   */
  public compile(): OpenAPIObject {
    return this.internalConfiguration.compile();
  }

  /**
   * Creates a Response object
   * @param httpCode The HTTP Code that the response represents
   * @returns A bagger response that can be used to create a compiled Swagger definition.
   * @example
   * ```js
   * const bagger = require('digitalroute/bagger');
   *
   * const getBags = bagger
   *   .response(200)
   *   .description('Successfully fetched all bags')
   *   .content('text/plain', { type: 'string' });
   * ```
   */
  public response(httpCode: number): BaggerResponse {
    return new BaggerResponse(httpCode);
  }

  public securityComponent(name: string, type: SecuritySchemeType): BaggerSecurityComponent {
    return new BaggerSecurityComponent(name, type);
  }

  /**
   * Create a request body used for defining a body in a bagger request
   * @example
   * ```js
   * const bagger = require('.');
   * const joi = require('joi');
   *
   * const body = bagger
   *   .requestBody()
   *   .description('Create a bag')
   *   .content('application/json',
   *     joi.object().keys({
   *       type: joi.string().valid('backpack', 'duffel', 'sports').required(),
   *       size: joi.array().items(joi.string().valid('10L', '20L', '30L').required()).required(),
   *       description: joi.string().optional()
   *     })
   *   )
   *   .required(true);
   *
   * bagger
   *   .addRequest('/bags', 'post')
   *   .body(body)
   * ...
   * ```
   */
  public requestBody(): BaggerRequestBody {
    return new BaggerRequestBody();
  }

  /**
   * Create a parameter used for defining query, path, cookie or header parameter in bagger requests.
   * @example
   * ```js
   * const bagger = require('bagger');
   * const parameter = bagger
   *   .parameter()
   *   .path('bagId')
   *   .schema(joi.string().required())
   *   .description('ID of one bag')
   *   .explode(true)
   *   .required(true);
   *
   * bagger
   *   .addRequest('/bags/{bagId}', 'get')
   *   .addParameter(parameter)
   * ...
   * ```
   */
  public parameter(): { [key in ParameterType]: (name: string) => BaggerParameter } {
    return {
      query: (name: string) => new BaggerParameter('query', name),
      path: (name: string) => new BaggerParameter('path', name),
      cookie: (name: string) => new BaggerParameter('cookie', name),
      header: (name: string) => new BaggerParameter('header', name)
    };
  }
}

export const defaultInstance = new Bagger();
