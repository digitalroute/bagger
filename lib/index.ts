import { bagger } from './bagger';
import joi from 'joi';

export { bagger, joi };
