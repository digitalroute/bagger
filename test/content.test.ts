import { Content } from '../lib/content';
import joi from 'joi';

describe('Content objects', () => {
  test('Simple content object can be created and compiled', () => {
    const content = new Content();
    content.add('application/json', joi.string().example('foo'));
    expect(content.compile()).toMatchSnapshot();
  });
});
