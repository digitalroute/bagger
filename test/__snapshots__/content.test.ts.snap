// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content objects Simple content object can be created and compiled 1`] = `
{
  "application/json": {
    "schema": {
      "example": "foo",
      "type": "string",
    },
  },
}
`;
